File: benchmarks/binary_trees.go
package main

import (
	"fmt"
	"time"
)

type Tree struct {
	l, r *Tree
}

func build(depth int) *Tree {
	if depth == 0 {
		return &Tree{nil, nil}
	}
	return &Tree{build(depth - 1), build(depth - 1)}
}

func check(t *Tree) int {
	if t == nil {
		return 0
	}
	return 1 + check(t.l) + check(t.r)
}

func main() {
	t0 := time.Now()
	total := 0
	for i := 0; i < 50; i++ {
		t := build(16)
		total += check(t)
	}
	fmt.Println(total)
	fmt.Printf("bench_ns=%d\n", time.Since(t0).Nanoseconds())
}

File: benchmarks/binary_trees.rs
struct Tree {
    l: Option<Box<Tree>>,
    r: Option<Box<Tree>>,
}

fn build(depth: i32) -> Option<Box<Tree>> {
    if depth == 0 {
        return Some(Box::new(Tree { l: None, r: None }));
    }
    Some(Box::new(Tree {
        l: build(depth - 1),
        r: build(depth - 1),
    }))
}

fn check(t: &Option<Box<Tree>>) -> i32 {
    match t {
        None => 0,
        Some(node) => 1 + check(&node.l) + check(&node.r),
    }
}

fn main() {
    let t0 = std::time::Instant::now();
    let mut total = 0;
    for _ in 0..50 {
        let t = build(16);
        total += check(&t);
    }
    let duration = t0.elapsed();
    println!("{}", total);
    println!("bench_ns={}", duration.as_nanos());
}
